exercises/complex-numbers/complex_numbers_test.cpp

```diff
@@ -0,0 +1,372 @@
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Subtract_numbers_with_real_and_imaginary_part)
+{
+    const complex_numbers::Complex c1 {1.0, 2.0};
+    
+    const complex_numbers::Complex c2 {3.0, 4.0};
+    
+    const complex_numbers::Complex actual = c1 - c2;
+
+    const complex_numbers::Complex expected {-2.0, -2.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Multiply_purely_real_numbers)
+{
+    const complex_numbers::Complex c1 {1.0, 0.0};
+    
+    const complex_numbers::Complex c2 {2.0, 0.0};
+    
+    const complex_numbers::Complex actual = c1 * c2;
+
+    const complex_numbers::Complex expected {2.0, 0.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Multiply_purely_imaginary_numbers)
+{
+    const complex_numbers::Complex c1 {0.0, 1.0};
+    
+    const complex_numbers::Complex c2 {0.0, 2.0};
+    
+    const complex_numbers::Complex actual = c1 * c2;
+
+    const complex_numbers::Complex expected {-2.0, 0.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Multiply_numbers_with_real_and_imaginary_part)
+{
+    const complex_numbers::Complex c1 {1.0, 2.0};
+    
+    const complex_numbers::Complex c2 {3.0, 4.0};
+    
+    const complex_numbers::Complex actual = c1 * c2;
+
+    const complex_numbers::Complex expected {-5.0, 10.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Imaginary_unit)
+{
+    const complex_numbers::Complex c1 {0.0, 1.0};
+    
+    const complex_numbers::Complex c2 {0.0, 1.0};
+    
+    const complex_numbers::Complex actual = c1 * c2;
+
+    const complex_numbers::Complex expected {-1.0, 0.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Divide_purely_real_numbers)
+{
+    const complex_numbers::Complex c1 {1.0, 0.0};
+    
+    const complex_numbers::Complex c2 {2.0, 0.0};
+    
+    const complex_numbers::Complex actual = c1 / c2;
+
+    const complex_numbers::Complex expected {0.5, 0.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Divide_purely_imaginary_numbers)
+{
+    const complex_numbers::Complex c1 {0.0, 1.0};
+    
+    const complex_numbers::Complex c2 {0.0, 2.0};
+    
+    const complex_numbers::Complex actual = c1 / c2;
+
+    const complex_numbers::Complex expected {0.5, 0.0};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Divide_numbers_with_real_and_imaginary_part)
+{
+    const complex_numbers::Complex c1 {1.0, 2.0};
+    
+    const complex_numbers::Complex c2 {3.0, 4.0};
+    
+    const complex_numbers::Complex actual = c1 / c2;
+
+    const complex_numbers::Complex expected {0.44, 0.08};
+
+    BOOST_CHECK(expected == actual);
+}
+
+BOOST_AUTO_TEST_CASE(Absolute_value_of_a_positive_purely_real_number)
+{
+    const complex_numbers::Complex c {5.0, 0.0};
+    
+    const double actual = c.abs();
+
+    const double expected {5.0};
+
+    BOOST_CHECK(expected == actual);
```

Comment:
You need to be careful comparing doubles here.

https://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/testing_tools/extended_comparison/floating_point.html

Reply:
Yes that's not an easy topic: [Comparing Floating Point](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/)